build-data-capturing-gradle-samples/capture-quality-check-issues/gradle-quality-check-issues.gradle

```diff
@@ -10,100 +13,143 @@ if (!buildScanApi) {
-            String type = bugInstance.@type.text()
-            def error = bugInstance.breadthFirst().findAll { it.name() == 'SourceLine' }.sort(false) { it.parent().name() as SpotBugsParent }.first()
-            String startLine = error.@start.text()
-            String endLine = error.@end.text()
-            String lineNumber = endLine == startLine ? startLine : "$startLine-$endLine"
-            def absoluteFilePath = bugInstance.parent().Project.SrcDir.collect { it.text() }.find { it.endsWith(error.@sourcepath.text()) }
-            String relativePath = task.project.rootProject.relativePath(absoluteFilePath)
-            "${relativePath}:${lineNumber} \u2192 ${type}"
-        }.flatten() as List<String>
-    } else if (ReportingTask.SPOTBUGS.isTask(task)) {
-        valueName = 'Verification SpotBugs'
-        errors = report.BugInstance.collect { bugInstance ->
-            String type = bugInstance.@type.text()
-            def error = bugInstance.breadthFirst().findAll { it.name() == 'SourceLine' }.sort(false) { it.parent().name() as SpotBugsParent }.first()
-            String startLine = error.@start.text()
-            String endLine = error.@end.text()
-            String lineNumber = endLine == startLine ? startLine : "$startLine-$endLine"
-            def absoluteFilePath = bugInstance.parent().Project.SrcDir.collect { it.text() }.find { it.endsWith(error.@sourcepath.text()) }
-            String relativePath = task.project.rootProject.relativePath(absoluteFilePath)
-            "${relativePath}:${lineNumber} \u2192 ${type}"
-        }.flatten() as List<String>
-    }
-
-    errors.each { buildScan.value valueName, it }
 }
 
-static String appendIfMissing(String str, String suffix) {
-    str.endsWith(suffix) ? str : str + suffix
+// check if reporting tasks are present in the build
+def hasReportingTask = false
+gradle.taskGraph.whenReady { taskGraph ->
+    if(taskGraph.allTasks.findAll {
+        it.class.name.startsWith('org.gradle.api.plugins.quality.Checkstyle')
+            || it.class.name.startsWith('org.gradle.api.plugins.quality.CodeNarc')
+            || it.class.name.startsWith('org.gradle.api.plugins.quality.FindBugs')
+            || it.class.name.startsWith('com.github.spotbugs.SpotBugsTask')
+    }.size() > 0) {
+        hasReportingTask = true
+    }
 }
 
-enum SpotBugsParent {
-
-    BugInstance, Method, Class
+// process results on build completion
+buildScan.buildFinished { def result ->
+    if(hasReportingTask) {
+        qualityCheckServiceProvider.get().processResults(buildScanApi)
+    }
 }
 
-enum ReportingTask {
-
-    CHECKSTYLE('org.gradle.api.plugins.quality.Checkstyle'),
-    CODENARC('org.gradle.api.plugins.quality.CodeNarc'),
-    FINDBUGS('org.gradle.api.plugins.quality.FindBugs'),
-    SPOTBUGS('com.github.spotbugs.SpotBugsTask')
+// Quality check service
+abstract class QualityCheckService implements BuildService<QualityCheckService.Params> {
```

Comment:
can you help me understand why we need the service here and we can't use a class like the other samples?

Reply:
Good point, I was under the impression that we were collecting/aggregating the results at the end of the build but it's happening just after each task.
Let me adjust the code.

Reply:
You can now review the service-less version of the _capture quality check issues sample_ @runningcode